contrib/internal/httptrace/inferred_proxy.go

```diff
@@ -0,0 +1,138 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package httptrace
+
+import (
+	"fmt"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"net/http"
+	"strconv"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/globalconfig"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+type StartSpanOption = ddtrace.StartSpanOption
+
+const (
+	ProxyHeaderSystem      = "X-Dd-Proxy"
+	ProxyHeaderStartTimeMs = "X-Dd-Proxy-Request-Time-Ms"
+	ProxyHeaderPath        = "X-Dd-Proxy-Path"
+	ProxyHeaderHttpMethod  = "X-Dd-Proxy-Httpmethod"
+	ProxyHeaderDomain      = "X-Dd-Proxy-Domain-Name"
+	ProxyHeaderStage       = "X-Dd-Proxy-Stage"
+)
+
+type ProxyDetails struct {
+	SpanName  string `json:"spanName"`
+	Component string `json:"component"`
+}
+
+var (
+	supportedProxies = map[string]ProxyDetails{
+		"aws-apigateway": {
+			SpanName:  "aws.apigateway",
+			Component: "aws-apigateway",
+		},
+	}
+)
+
+type ProxyContext struct {
+	RequestTime     string `json:"requestTime"`
+	Method          string `json:"method"`
+	Path            string `json:"path"`
+	Stage           string `json:"stage"`
+	DomainName      string `json:"domainName"`
+	ProxySystemName string `json:"proxySystemName"`
+}
+
+func extractInferredProxyContext(headers http.Header) *ProxyContext {
+	_, exists := headers[ProxyHeaderStartTimeMs]
+	if !exists {
+		log.Debug("Proxy header start time does not exist")
+		return nil
+	}
+
+	proxyHeaderSystem, exists := headers[ProxyHeaderSystem]
+	if !exists {
+		log.Debug("Proxy header system does not exist")
+		return nil
+	}
+
+	if _, ok := supportedProxies[proxyHeaderSystem[0]]; !ok {
+		log.Debug("Unsupported Proxy header system")
+		return nil
+	}
+
+	return &ProxyContext{
+		RequestTime:     headers[ProxyHeaderStartTimeMs][0],
```

Comment:
```suggestion
		RequestTime:    headers.Get(ProxyHeaderStartTimeMs),
```

(same for everywhere else where you read headers)

also, you could directly do the string parsing to unix mill here in this function:

```
        startTimeUnixMilli, err := strconv.ParseInt(headers.Get(ProxyHeaderStartTimeMs), 10, 64)
	if err != nil {
		log.Debug("Error parsing time string: %v", err)
		return nil, fmt.Errorf("failed to parse start time from headers: %w", err)
	}
	startTime := time.UnixMilli(startTimeUnixMilli)
```